### PR TITLE
convert cloud hosted custom module into a policy

### PR DESCRIPTION
**Pull request by mikz**
This is an attempt to use https://github.com/3scale/apicast/pull/450 in production use case.

## TODO:
- [x] no more rockspec - use roverfile
- [x] move sources to correct folder (because rockspec won't copy them)
- [x] fix `/usr/bin/container-entrypoint: line 16: /usr/libexec/s2i/run: No such file or directory` when running built image without a command (https://github.com/3scale/s2i-openresty/pull/42)
- [x] convert custom module into policies (https://github.com/3scale/apicast/issues/481)
- [x] configure APIcast to use those policies

### CODE REVIEW

**Comment by mikz** [comment]
@3scale/operations we need to figure out how to use stock apicast image and add this customization with s2i to build new image on OpenShift.

File: apicast/Makefile [hunk 1]
```diff
@@ -1,6 +1,6 @@
 .DEFAULT_GOAL := help
 
-APICAST_VERSION ?= v3.0.0
+APICAST_VERSION ?= policy-chain
```
Comment (davidor) [inline apicast/Makefile]: I don't see this version in quay.

Comment (davidor) [reply apicast/Makefile]: Ah forget about it, I see that you changed this in another commit.

File: apicast/cpanfile [hunk 2]
```diff
@@ -0,0 +1 @@
+requires 'Test::APIcast', '0.04';
```
Comment (davidor) [inline apicast/cpanfile:1]: `0.04` is an old version. Although the latest will be installed anyway unless we pin the version with `==`.

File: .circleci/config.yml [hunk 3]
```diff
@@ -1,26 +1,79 @@
...
             if [ -n "${CIRCLE_TAG}" ] || [ -n "${CIRCLE_BRANCH}" ]; then
-              docker login -u="${DOCKER_USER}" -p="${DOCKER_PASS}" "${DOCKER_REGISTRY}"
+              docker login -u="${DOCKER_USERNAME}" --password-stdin "${DOCKER_REGISTRY}" <<< "${DOCKER_PASSWORD}"
               (cd apicast && make push REMOTE_IMAGE_NAME=apicast-cloud-hosted:apicast-${CIRCLE_TAG:-${CIRCLE_BRANCH}})
               (cd mapping-service && make push IMAGE_TAG=${CIRCLE_TAG:-${CIRCLE_BRANCH}})
             fi
 
+  apicast-test:
+    docker:
+      - image: quay.io/3scale/s2i-openresty-centos7:1.13.6.1-rover5
```
Comment (davidor) [inline .circleci/config.yml]: Latest is `s2i-openresty-centos7:1.13.6.1-rover6`

File: apicast/t/blacklist.t [hunk 4]
```diff
@@ -0,0 +1,68 @@
+BEGIN {
+    $ENV{TEST_NGINX_APICAST_BINARY} ||= 'rover exec apicast';
+    $ENV{APICAST_POLICY_LOAD_PATH} = './policies';
+}
+
+use strict;
+use warnings FATAL => 'all';
+use Test::APIcast::Blackbox 'no_plan';
+
+repeat_each(1);
```
Comment (davidor) [inline apicast/t/blacklist.t]: I think this one can be run twice. Delete this line so Test::APIcast applies its default?

File: apicast/t/module.t [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    $ENV{TEST_NGINX_APICAST_BINARY} ||= 'APICAST_MODULE="" rover exec ./lua_modules/bin/apicast';
+    $ENV{APICAST_MODULE} = 'cloud_hosted.module';
+    $ENV{APICAST_POLICY_LOAD_PATH} = './policies';
+}
+
+use strict;
+use warnings FATAL => 'all';
+use Test::APIcast::Blackbox 'no_plan';
+
+repeat_each(1);
```
Comment (davidor) [inline apicast/t/module.t]: Same.

File: apicast/cpanfile (hunk 2 above)
Comment (mikz) [reply apicast/cpanfile:1]: Yep. I think that is fine. This library is already installed in the apicast image so this just ensures it has the right version.

**Pull request merged**